tests/test_solvers.py

```diff
@@ -0,0 +1,295 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+import collections
+import contextlib
+import json
+import os
+import pathlib
+import re
+import tempfile
+import textwrap
+
+from typing import Type
+
+import pytest
+
+import conda.core.solve
+
+from conda.base.context import context
+from conda.base.constants import REPODATA_FN
+from conda.core.subdir_data import SubdirData
+from conda.exceptions import ResolvePackageNotFound, UnsatisfiableError
+from conda.models.channel import Channel
+from conda.resolve import MatchSpec
+
+from . import helpers
+
+
+class SolverTests:
+    """Tests for :py:class:`conda.core.solve.Solver` implementations."""
+
+    @property
+    def solver_class(self) -> Type[conda.core.solve.Solver]:
+        """Class under test."""
+        raise NotImplementedError
+
+    def _write_channel_packages(self, channel_path: pathlib.Path, subdirs, packages):
+        """Create a channel containing a set of packages."""
+        # build package data
+        package_data = collections.defaultdict(dict)
+        import conda.models.records
+        for record in packages:
+            assert isinstance(record, conda.models.records.PackageRecord)
+            package_data[record.subdir][record.fn] = {
+                key: value
+                for key, value in vars(record).items()
+                if key in (
+                    'build',
+                    'build_number',
+                    'depends',
+                    'license',
+                    'md5',
+                    'name',
+                    'sha256',
+                    'size',
+                    'subdir',
+                    'timestamp',
+                    'version',
+                    'track_features',
+                    'features',
+                )
+            }
+        # write repodata
+        for subdir in subdirs:
+            subdir_path = channel_path / subdir
+            subdir_path.mkdir(parents=True, exist_ok=True)
+            subdir_path.joinpath('repodata.json').write_text(json.dumps({
+                'info': {
+                    'subdir': subdir,
+                },
+                'packages': package_data.get(subdir, {}),
+            }))
+
+    @contextlib.contextmanager
+    def env(self, key, value):
+        old = os.environ.get(key)
+        os.environ[key] = value
+        try:
+            yield
+        except:
+            if old is not None:
+                os.environ[key] = old
```

Comment:
Is there a chance that restoring the environment could fail if something happens in the context manager's context, I've usually used pytest's request.addfinalizer for this and I'm worried we'd add a footgun with this.

Reply:
There shouldn't. Using `request.addfinalizer` would complicate the API, if we are doing that we should probably use https://docs.pytest.org/en/stable/monkeypatch.html instead. But I don't think I am using this helper anymore, so we can just remove it.

Reply:
Gotcha, good thinking!